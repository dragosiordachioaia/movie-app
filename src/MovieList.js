import React, { Component } from "react";

import MovieItem from "./MovieItem";

export default class MovieList extends Component {
  constructor(props) {
    super(props);

    this.displayList = this.displayList.bind(this);
  }

  displayList() {
    return this.props.movies.map(movieData => {
      return (
        <div className="col-sm-3 col-xs-6" key={movieData.id}>
          <MovieItem data={movieData} config={this.props.config} />
        </div>
      );
    });
  }

  render() {
    console.log("movies are:", this.props.movies);
    return <div> {this.displayList()} </div>;
  }
}
