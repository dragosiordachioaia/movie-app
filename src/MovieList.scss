@import "variables";

.movie-list-container {
  background-color: #232b31;
  overflow: auto;
  padding: 2em 0.25em 1em;
  height: calc(100vh - #{$header-height} - #{$sidebar-height});
  width: 100%;
  transition: height 500ms;
  transition-timing-function: ease-in-out;
  position: relative;

  @media (min-width: $SM) {
    height: 100%;
    width: calc(100% - #{$sidebar-width});
    padding: 1em 0.5em;
  }

  @media (min-width: $MD) {
    padding: 2em 0.5em;
  }

  & > .title {
    color: #fff;
    margin-top: 0;
    font-weight: 100;
    margin-left: 1rem;
    margin-bottom: 1em;
    letter-spacing: 1px;
  }

  .movie-list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
  }
}
