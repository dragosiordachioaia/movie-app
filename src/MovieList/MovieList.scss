@import "../variables";

.movie-list-container {
  background-color: #232b31;
  overflow: auto;
  padding: 1em 0.25em;
  height: calc(100vh - #{$header-height} - #{$sidebar-height});
  width: 100%;
  transition: height 500ms;
  transition-timing-function: ease-in-out;
  position: relative;
  -webkit-overflow-scrolling: touch;

  @media (min-width: $SM) {
    height: 100%;
    width: calc(100% - #{$sidebar-width});
    padding: 1em 0.5em;
  }

  @media (min-width: $MD) {
    padding: 2em 0.5em;
  }

  & > .title {
    color: #fff;
    margin-top: 0;
    font-weight: 100;
    margin-left: 0.5rem;
    margin-bottom: 2em;
    letter-spacing: 1px;
    font-weight: 700;
    font-size: 1.3em;

    @media (min-width: $SM) {
      font-weight: 400;
      font-size: 1.9em;
      margin-bottom: 1.2em;
    }
  }

  .movie-list {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
  }
}
