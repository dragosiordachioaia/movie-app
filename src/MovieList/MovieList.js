import React, { Component } from "react";

import "./MovieList.scss";

import MovieItem from "MovieItem/MovieItem";

export default class MovieList extends Component {
  constructor(props) {
    super(props);

    this.displayMovies = this.displayMovies.bind(this);
    this.getFilteredList = this.getFilteredList.bind(this);
  }

  displayMovies(list) {
    return list.sort((a, b) => b.popularity - a.popularity).map(movieData => {
      return (
        <MovieItem
          key={movieData.id}
          data={movieData}
          config={this.props.config}
          genres={this.props.genres}
        />
      );
    });
  }

  getFilteredList() {
    return this.props.movies.filter(movieData => {
      let ratingIsOk = movieData.vote_average >= this.props.minRating;
      let genreIsOk = true;
      let nameIsOk = true;
      let selectedGenres = this.props.genres.filter(
        genreData => genreData.checked
      );
      if (this.props.nameFilter.length > 0) {
        nameIsOk = movieData.title
          .toLowerCase()
          .includes(this.props.nameFilter.toLowerCase());
      }
      if (selectedGenres.length > 0) {
        selectedGenres.forEach(({ id }) => {
          if (!movieData.genre_ids.includes(id)) {
            genreIsOk = false;
          }
        });
      }
      return genreIsOk && ratingIsOk && nameIsOk;
    });
  }

  render() {
    let filteredList = this.getFilteredList();
    let title = "Films showing right now";
    if (filteredList.length === 0) {
      title = "No movies match your search";
    }
    return (
      <div className="movie-list-container">
        <h1 className="title">{title}</h1>
        <div className="movie-list">{this.displayMovies(filteredList)}</div>
      </div>
    );
  }
}
