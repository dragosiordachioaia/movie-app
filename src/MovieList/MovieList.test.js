import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import MovieList from "./MovieList";

let component;
let instance;

beforeEach(() => {
  component = renderer.create(<MovieList movies={[]} />);
  instance = component.getInstance();
});

it("renders correctly when no movies", () => {
  instance.getFilteredList = jest.fn().mockReturnValue([]);
  instance.displayMovies = jest.fn().mockReturnValue("__movies-here__");
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("renders correctly when it has movies", () => {
  instance.props = { movies: [{}, {}], genres: [], nameFilter: "" };
  instance.getFilteredList = jest.fn().mockReturnValue([{}, {}]);
  instance.displayMovies = jest.fn().mockReturnValue("__movies-here__");
  const result = instance.render();
  expect(result).toMatchSnapshot();
});

it("correctly displays and sorts movies", () => {
  instance.props = { movies: [{}, {}], genres: [], nameFilter: "" };
  let list = [
    {
      popularity: 0,
      id: 0,
    },
    {
      popularity: 1,
      id: 1,
    },
  ];
  const result = instance.displayMovies(list);
  expect(result).toMatchSnapshot();
});

it("correctly filters movies", () => {
  instance.props = {
    movies: [
      {
        popularity: 2,
        title: "Avengers 1",
        genre_ids: [0, 1],
        vote_average: 5,
      },
      {
        popularity: 3,
        title: "Avengers 2",
        genre_ids: [0, 1],
        vote_average: 4,
      },
      {
        popularity: 4,
        title: "Avengers 3",
        genre_ids: [0, 1],
        vote_average: 6,
      },
      {
        popularity: 5,
        title: "Venom",
        genre_ids: [0],
        vote_average: 7,
      },
    ],
    genres: [
      {
        id: 0,
        name: "Action",
        checked: true,
      },
      {
        id: 1,
        name: "Adventure",
        checked: true,
      },
      {
        id: 2,
        name: "Comedy",
        checked: false,
      },
    ],
    nameFilter: "av",
    minRating: 5,
  };
  const result = instance.getFilteredList();
  expect(result).toMatchSnapshot();
});
