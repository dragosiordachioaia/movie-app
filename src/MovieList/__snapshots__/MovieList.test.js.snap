// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly displays and sorts movies 1`] = `
Array [
  <MovieItem
    data={
      Object {
        "id": 1,
        "popularity": 1,
      }
    }
    genres={Array []}
  />,
  <MovieItem
    data={
      Object {
        "id": 0,
        "popularity": 0,
      }
    }
    genres={Array []}
  />,
]
`;

exports[`correctly filters movies 1`] = `
Array [
  Object {
    "genre_ids": Array [
      0,
      1,
    ],
    "popularity": 2,
    "title": "Avengers 1",
    "vote_average": 5,
  },
  Object {
    "genre_ids": Array [
      0,
      1,
    ],
    "popularity": 4,
    "title": "Avengers 3",
    "vote_average": 6,
  },
]
`;

exports[`renders correctly when it has movies 1`] = `
<div
  className="movie-list-container"
>
  <h1
    className="title"
  >
    Movies showing right now
  </h1>
  <div
    className="movie-list"
  >
    __movies-here__
  </div>
</div>
`;

exports[`renders correctly when no movies 1`] = `
<div
  className="movie-list-container"
>
  <h1
    className="title"
  >
    No movies match your search
  </h1>
  <div
    className="movie-list"
  >
    __movies-here__
  </div>
</div>
`;
